docs/proposals/failover/optimize-cluster-failover.md

```diff
@@ -0,0 +1,635 @@
+
+![cluster failover overview](optimize-cluster-failover-overview.png)
+
+Regarding the overall process of Cluster Failover, after re-sorting, it is divided into the following stages:
+1. Input of cluster status conditions.
+2. Matching of cluster failure strategies.
+3. System response to failures (preventing the scheduling of new workloads or triggering workload failure migration).
+
+### Overall Design Principles
+
+Below, we present the overall design principles of Cluster Failover:
+
+1. Judgment of cluster failure conditions
+- The sources for judging cluster failures are diverse. The cluster status detection that Karmada supports by default will first be reflected in `cluster.status.conditions`.
+- After the Karmada system determines that a cluster has a failure, it will add corresponding taints to the cluster object. In cases of severe failures, the migration process will be triggered.
+- Users can define "which taints should be added to the cluster object when certain conditions occur" by configuring the "taint cluster by conditions" rule.
+- For more complex scenarios of cluster failure judgment, users can directly implement a custom "cluster taint controller" to control how to add and remove taints from the cluster.
+
+2. Reconsider the types of taints and adjust them into the following categories according to the severity of cluster failures and the scope of affected workloads:
+- Impact new workloads' scheduling and running: taint{effect: NoSchedule}. New workloads will not be scheduled, and will not trigger resource migration.
+- Impact all workloads' scheduling and running: taint{effect: NoExecute}(Reserve semantics: Hard migration, requiring users to add relevant comments to remind them of the impact in the API definition). New workloads will not be scheduled, and existing workloads will be migrated.
+- Add a new type of taint between taint{effect: NoSchedule} and taint{effect: NoExecute}: taint{effect: PreferNoExecute}.
+    - Default to not migrating, no need to consider tolerations (defined in `PP/CPP.spec.placement.clusterTolerations`).
+    - Active migration strategy, which can be configured through `PP/CPP.spec.failover.cluster`.
+
+3. Application-level cluster failover strategy:
+- Reinstate: Hard migration for taint{effect: NoExecute}.
+    - Migration method: graceful or directly (need to be decided).
+- New addition: Respond to taint{effect: PreferNoExecute}.
+- The failure migration strategy at the workload granularity needs to include the following elements: :
+    - Migration method: Graciously, Directly
+    - Toleration time
+
+4. Eviction queue (The queue maintained in memory)
+- It is necessary to consider the overall health of the Karmada and have the ability to downgrade eviction behavior and slow down queue processing.
+- During the eviction process, that is, when there are workloads to be evicted in the queue and the original faulty cluster recovers, the migration of these workloads can be abandoned.
```

Comment:
I don't think so.
How do you `cancel` the migration? Is there a safe way provided for users to manually cancel a migration?

Reply:
Same question here - 
what happens if a resource is halfway through migration when recovery happens? Does the system rollback partially migrated resources, or just leave them in the new location?

Reply:
Yes, that's my concern too.

Reply:
At this time, the evicting action has not occurred, and the new scheduling result has not been generated. Therefore, if the evicting action is canceled, the resource distribution will not be modified.

Reply:
Once the migration behavior has occurred, it cannot be rolled back, refer to here: https://github.com/karmada-io/karmada/pull/6274/files#diff-4c1fbc67529e131bae53ba0962bc08c2a158fa392700d024741e67f024941f5dR513